dd-sdk-android/src/main/cpp/native-lib.cpp

```diff
@@ -0,0 +1,194 @@
+//
+// Native Library for Exceptions logging.
+//
+
+// Java
+#include <jni.h>
+// C++
+#include <csignal>
+#include <cstdio>
+#include <cstring>
+#include <exception>
+#include <memory>
+// C++ ABI
+#include <cxxabi.h>
+// Android
+#include <android/log.h>
+#include <unistd.h>
+
+/// Helper macro to get size of an fixed length array during compile time
+#define sizeofa(array) sizeof(array) / sizeof(array[0])
+
+/// tgkill syscall id for backward compatibility (more signals available in many linux kernels)
+#define __NR_tgkill 270
+
+/// Caught signals
+static const int SIGNALS_TO_CATCH[] = {
+        SIGABRT,
+        SIGBUS,
+        SIGFPE,
+        SIGSEGV,
+        SIGILL,
+        SIGSTKFLT,
+        SIGTRAP,
+};
+/// Signal handler context
+struct CrashInContext {
+    /// Old handlers of signals that we restore on de-initialization. Keep values for all possible
+    /// signals, for unused signals nullptr value is stored.
+    struct sigaction old_handlers[NSIG];
+};
+/// Crash handler function signature
+typedef void (*CrashSignalHandler)(int, siginfo*, void*);
+/// Global instance of context. Since an app can't crash twice in a single run, we can make this singleton.
+static CrashInContext* crashInContext = nullptr;
+
+/// Register signal handler for crashes
+static bool registerSignalHandler(CrashSignalHandler handler, struct sigaction old_handlers[NSIG]) {
+    struct sigaction sigactionstruct;
+    memset(&sigactionstruct, 0, sizeof(sigactionstruct));
+    sigactionstruct.sa_flags = SA_SIGINFO;
+    sigactionstruct.sa_sigaction = handler;
+    // Register new handlers for all signals
+    for (int index = 0; index < sizeofa(SIGNALS_TO_CATCH); ++index) {
+        const int signo = SIGNALS_TO_CATCH[index];
+        if (sigaction(signo, &sigactionstruct, &old_handlers[signo])) {
+            return false;
+        }
+    }
```

Comment:
I'm wondering if it's the same in Android NDK, but the [PLCrashReporter](https://github.com/microsoft/plcrashreporter) (the widely used crash reporter on iOS) has [the limitation of registering only one signal handler](https://github.com/microsoft/plcrashreporter/blob/748087386cfc517936315c107f722b146b0ad1ab/Source/PLCrashSignalHandler.mm#L359-L360) (this is in fact OS limitation).

The question from here is: can our NDK Crash reporter work together with other 3rd party reporters, or it will disable them / be disabled? If we're not sure, it's maybe worth testing.

Reply:
That's a very good question, as for me this NDK world is totally unknown now I am not sure what it will do. I will study more the C++ part of the code and then maybe try to register 2 handlers and see if they both work in the same time.